### PR TITLE
server: use an EnumSetting for server.remote_debugging.mode

### LINKED ISSUE #15602: server: make cluster setting "server.remote_debugging.mode" into an enum setting
Now that we have enum settings in #15307, this setting should be an enum setting, as it currently fails silently if the value is not one of "any", "true", "t", or "1".

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #15602

### CODE REVIEW

File: pkg/server/debug.go [hunk 1]
```diff
@@ -41,11 +40,22 @@ const debugEndpoint = "/debug/"
 // register to that via import, and go-metrics registers to that via exp.Exp())
 var debugServeMux = http.DefaultServeMux
 
-// TODO(arjun): use an Enum setting here.
-var debugRemote = settings.RegisterStringSetting(
```
Comment (dt) [inline pkg/server/debug.go]: Going forward, in-place type changes are usually a Bad Idea as they'll make deserialization angry, but probably fine pre-1.0.

**Review by dt — APPROVED** [review]

**Comment by arjunravinarayan** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/debug.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-KjFQSfNrrkOaqNOOF_h:-KjFQSfNrrkOaqNOOF_i:bswp41d) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f8f2fd88f164c304a337f69577cdc12e9b92aa/pkg/server/debug.go#L78)):*
> ```Go
> 	trace.AuthRequest = func(req *http.Request) (bool, bool) {
> 		allow, sensitive := origAuthRequest(req)
> 		switch debugRemote.Get() {
> ```

How about
```
switch debugRemote.Get() {
case debugRemoteAny:
   return true, sensitive
case debugRemoteLocal:
   return allow, sensitive
default:
   return false, sensitive
}
```
Easier on the eyes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjFQw7bKiFwF1a8akHn:bnfp4nl)*

**Comment by arjunravinarayan** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/debug.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-KjFQSfNrrkOaqNOOF_h:-KjFS0__h_P6NACYuRrL:b-6ev2di) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f8f2fd88f164c304a337f69577cdc12e9b92aa/pkg/server/debug.go#L78)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

How about
```
switch debugRemote.Get() {
case debugRemoteAny:
   return true, sensitive
case debugRemoteLocal:
   return allow, sensitive
default:
   return false, sensitive
}
```
Easier on the eyes.
</blockquote></details>

Not your code, and not something I feel strongly about, so don't bother fixing it. It was confusing me what `sensitive` was doing given that we don't really touch it at all or use it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678)*

**Comment by petermattis** [comment]
@dt 

> Going forward, in-place type changes are usually a Bad Idea as they'll make deserialization angry, but probably fine pre-1.0.

This will be post-1.0, but not that `string` and `enum` deserialization consume the same type (I think).

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/debug.go, line 78 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-KjFQSfNrrkOaqNOOF_h:-KjFrfMkamQeMr7jPQT2:b6dt4l3) ([raw file](https://github.com/cockroachdb/cockroach/blob/91f8f2fd88f164c304a337f69577cdc12e9b92aa/pkg/server/debug.go#L78)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Not your code, and not something I feel strongly about, so don't bother fixing it. It was confusing me what `sensitive` was doing given that we don't really touch it at all or use it.
</blockquote></details>

Gah. This actually is my code. I was targeting the minimal diff here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjFrq9TJGuJpGWb2xlA:bqjpn8r)*

**Comment by petermattis** [comment]
`s/not/note/g`

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjFsXvRbKqeTrc0eyS9:b-o4uiu5)*

**Comment by dt** [comment]
enums are stored as ints -- @knz wanted this to allow renaming their possible values -- so i think deserialization will object.

That said, I don't think we'll even attempt to deserialize it, since the `valueType` field for an enum is `e` but `s` for a string, and that is used to sanity check and reject values with mis-matched types when parsing the gossip update.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjGC5VcGna7cNLcJPnv:b-ozqfea)*

**Comment by petermattis** [comment]
Got it. Switched back to a string, but with a validation function specified. PTAL.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjIRR2vM6Cq0LMAQGUd:b-p74s5r)*

**Comment by dt** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15678#-:-KjIW1Isaxu1KZDN0NWZ:bnfp4nl)*

**Comment by knz** [comment]
I don't feel good about using a string setting with a validation function here, since we implemented EnumSetting.

This is a case where I would recommend we simply ignore the backward compatibility and instead go for an EnumSetting right away. As David says the gossip code will ignore existing configurations because the types don't match; I think we should not care since we are probably the only people using this setting for now anyway.

**Comment by petermattis** [comment]
> I don't feel good about using a string setting with a validation function here, since we implemented EnumSetting.

I don't feel strongly one way or another, but can you elaborate on why a validation function is problematic. Seemed elegant to me on the surface.

**Comment by knz** [comment]
The issue is that the EnumSetting code already does the same thing. This looks like code redundancy and DRY violation.

**Comment by petermattis** [comment]
> The issue is that the EnumSetting code already does the same thing. This looks like code redundancy and DRY violation.

That doesn't bother me too much, but I realize I'm exceptional in that way. I could also make this more elegant (`RegisterRestrictedStringSet`). I'll wait for @dt to weigh in.

**Comment by dt** [comment]
Heh, I don't have particularly strong opinions and am fine with either the previous approach (swap, with a release note) or the current code.

Mildly opposed to adding another setting variant (restricted string) for (imo) very minimal DRY benefit: future users probably want an enum and if not, they probably need the control offered by an arbitrary func anyway, and this validation func is pretty as-is clear to me.

**Comment by knz** [comment]
Ok I just stopped caring as well, so now there's 3 of us who don't care.

Would be nice to promote enum settings more in the future though.

**Comment by petermattis** [comment]
Ok, I'm going to merge as-is, because I'm conservative about backwards compatibility.

**Pull request merged**